tools/k8s_versions/update_k8s_versions.go

```diff
@@ -0,0 +1,287 @@
+	if err = json.Unmarshal(content, &testMatrix); err != nil {
+		return fmt.Errorf("failed to unmarshal JSON: %w", err)
+	}
+
+	for _, value := range testMatrix {
+		if len(kindVersions) > 0 && value[KubeKindVersion] != nil {
+			value[KubeKindVersion] = kindVersions
+		} else if len(minikubeVersions) > 0 && value[KubeMinikubeVersion] != nil {
+			value[KubeMinikubeVersion] = minikubeVersions
+		}
+	}
+	// Marshal the updated test matrix back to JSON
+	updatedContent, err := json.MarshalIndent(testMatrix, "", "  ")
+	if err != nil {
+		return fmt.Errorf("failed to marshal updated JSON: %w", err)
+	}
+
+	// Ensure the file ends with a new line to make the pre-commit check happy
+	updatedContent = append(updatedContent, '\n')
+
+	if err = os.WriteFile(filePath, updatedContent, 0644); err != nil {
+		return fmt.Errorf("failed to write updated file: %w", err)
+	}
+	return nil
+}
+
+func sortVersions(versions []string) {
+	sort.Slice(versions, func(i, j int) bool {
+		vi := strings.Split(versions[i][1:], ".") // Remove "v" and split by "."
+		vj := strings.Split(versions[j][1:], ".")
+
+		for k := 0; k < len(vi) && k < len(vj); k++ {
+			if vi[k] != vj[k] {
+				return vi[k] > vj[k] // Sort in descending order
+			}
+		}
+		return len(vi) > len(vj)
+	})
+}
+
+func getRequestBody(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch URL: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+	return body, nil
+}
+
+func logDebug(format string, v ...interface{}) {
+	if debug {
+		log.Printf(format, v...)
+	}
+}
+
+func main() {
+	// setup logging
+	flag.BoolVar(&debug, "debug", false, "Enable debug logging")
+	flag.Parse()
+	log.SetOutput(os.Stdout)
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	k8sVersions, err := getSupportedKubernetesVersions(EndOfLifeURL)
+	if err != nil || len(k8sVersions) == 0 {
+		log.Fatalf("Failed to get k8s versions: %v", err)
+	}
+	logDebug("Found supported k8s versions %v", k8sVersions)
+
+	kindVersions, err := getLatestSupportedKindImages(KindDockerHubURL, k8sVersions)
+	if err != nil {
+		log.Printf("failed to get all kind versions: %v", err)
+	}
+	if len(kindVersions) > 0 {
+		// needs to be sorted so we don't end up with false positive diff in the json matrix file
+		sortVersions(kindVersions)
+		logDebug("Found supported kind images: %v", kindVersions)
+	}
+
+	minikubeVersions, err := getLatestSupportedMinikubeVersions(MiniKubeURL, k8sVersions)
+	if err != nil {
+		log.Printf("failed to get minikube versions: %v", err)
+	}
+	if len(minikubeVersions) > 0 {
+		logDebug("Found supported minikube versions: %v", minikubeVersions)
+	}
+
+	if len(kindVersions) == 0 && len(minikubeVersions) == 0 {
```

Comment:
Shouldn't we fail if any of them is empty not both?

Reply:
I figured if one provider failed (ex: the endpoint was not reachable), we should not sacrifice the other